### PR TITLE
implements display based subtype checking (6.4x faster without threads; 2.8x faster with threads)

### PR DESCRIPTION
**Pull request by ringabout**
ref https://github.com/nim-lang/Nim/pull/20777

todo
- [x] cpp
- [x] downcast
- [x] benchmarks

Here is the benchmark https://github.com/nim-lang/Nim/issues/18612#issuecomment-890531425

### CODE REVIEW

File: lib/system/arc.nim [hunk 1]
```diff
@@ -256,6 +256,12 @@ proc isObjWithCache(obj: PNimTypeV2, subclass: cstring, cache: var ObjCheckCache
   else:
     result = isObjSlowPath(obj, subclass, cache)
 
+proc isObjDisplayCheck(source: PNimTypeV2, target: PNimTypeV2, token: uint32): bool {.compilerRtl.} =
```
Comment (Araq) [inline lib/system/arc.nim]:
```suggestion
proc isObjDisplayCheck(source: PNimTypeV2, targetDepth: int16, token: uint32): bool {.compilerRtl.} =
```

The compiler can pass `targetDepth` directly here as it knows it. Saves a memory fetch operation.

Comment (ringabout) [reply lib/system/arc.nim]: done

**Comment by Araq** [comment]
I declare this to be crucial for version 2.

File: lib/system/arc.nim [hunk 2]
```diff
@@ -242,19 +240,11 @@ func endsWith(s, suffix: cstring): bool {.inline.} =
...
+proc isObjDisplayCheck(source: PNimTypeV2, targetDepth: int16, token: uint32): bool {.compilerRtl.} =
+  if targetDepth > source.depth:
     result = false
-
-proc isObjWithCache(obj: PNimTypeV2, subclass: cstring, cache: var ObjCheckCache): bool {.compilerRtl.} =
-  if cache[0] == obj: result = false
-  elif cache[1] == obj: result = true
   else:
-    result = isObjSlowPath(obj, subclass, cache)
+    result = source.display[targetDepth] == token
```
Comment (Araq) [inline lib/system/arc.nim]:
```suggestion
    result = targetDepth <= source.depth and source.display[targetDepth] == token
```

Comment (ringabout) [reply lib/system/arc.nim]: done

**Comment by Araq** [comment]
CI failure unrelated, merging.

**Comment by rockcavera** [comment]
This PR broke compiler compilation here.

```
e:\Nim>koch boot -d:release --skipUserCfg --skipParentCfg
iteration: 1
bin\nim.exe c  --skipUserCfg --skipParentCfg -d:nimKochBootstrap --nimcache:nimcache/r_windows_amd64 -d:release --skipUserCfg --skipParentCfg --noNimblePath --compileOnly compiler\nim.nim
Hint: used config file 'e:\Nim\config\nim.cfg' [Conf]
Hint: used config file 'e:\Nim\config\config.nims' [Conf]
Hint: used config file 'e:\Nim\compiler\nim.cfg' [Conf]
...............................................................................................................................................................................................................................
e:\Nim\compiler\ccgutils.nim(56, 14) Warning: target type is larger than source type
  target type: 'Hash' (8 bytes)
  source type: 'uint32' (4 bytes) [CastSizes]
................
e:\Nim\compiler\vmops.nim(149, 8) Warning: gc is deprecated [Deprecated]
..................................................................
Error: system module needs: isObjWithCache
FAILURE
```

```
nim -v
Nim Compiler Version 1.7.3 [Windows: amd64]
Compiled at 2022-11-08
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: ca3b6cba5dca1b0f73d8dec7476d440af9780567
active boot switches: -d:release
```

**Comment by ringabout** [comment]
Hi @rockcavera
How about clearing the nimcache and removing the old `bin/nim` executable?

**Comment by rockcavera** [comment]
@ringabout , It worked! Thanks.

**Comment by ringabout** [comment]
No problem.

**Comment by rockcavera** [comment]
Looks like this PR caused a regression in performance using --mm:refc and --threads:on

```
REFC

before this PR -d:release --mm:refc --threads:off
 Proc with ref object 0.107
 Closures 2.246
 Methods 0.3599999999999999

after this PR -d:release --mm:refc --threads:off
 Proc with ref object 0.108
 Closures 2.092
 Methods 0.3260000000000001

before this PR -d:release --mm:refc --threads:on
 Proc with ref object 0.107
 Closures 4.382000000000001
 Methods 0.4370000000000003

after this PR -d:release --mm:refc --threads:on
 Proc with ref object 0.107
 Closures 11.914
 Methods 1.298999999999999


ARC

before this PR -d:release --mm:arc --threads:off
 Proc with ref object 0.107
 Closures 1.78
 Methods 2.605

after this PR -d:release --mm:arc --threads:off
 Proc with ref object 0.108
 Closures 1.645
 Methods 0.3240000000000003

before this PR -d:release --mm:arc --threads:on
 Proc with ref object 0.108
 Closures 7.931
 Methods 4.205

after this PR -d:release --mm:arc --threads:on
 Proc with ref object 0.108
 Closures 7.888
 Methods 1.329999999999999

before this PR -d:release --mm:arc --threads:on -d:useMalloc
 Proc with ref object 0.107
 Closures 13.06
 Methods 3.424999999999999

after this PR -d:release --mm:arc --threads:on -d:useMalloc
 Proc with ref object 0.108
 Closures 14.402
 Methods 1.850999999999999

before this PR -d:release --mm:arc --threads:off -d:useMalloc
 Proc with ref object 0.108
 Closures 9.478
 Methods 2.664999999999999

after this PR -d:release --mm:arc --threads:off -d:useMalloc
 Proc with ref object 0.106
 Closures 9.887
 Methods 0.3569999999999993
```

**Comment by ringabout** [comment]
I cannot reproduce on Win11

`bin/nim r -d:release --mm:refc --threads:off `
```
 Proc with ref object 0.045
 Closures 1.111
 Methods 0.1920000000000002
```

`bin/nim r -d:release --mm:refc --threads:on`
```
 Proc with ref object 0.046
 Closures 2.146
 Methods 0.2000000000000002
```

**Comment by rockcavera** [comment]
I talk about a comparison between performance before PR and after PR:
```
before this PR -d:release --mm:refc --threads:on
 Proc with ref object 0.107
 Closures 4.382000000000001
 Methods 0.4370000000000003

after this PR -d:release --mm:refc --threads:on
 Proc with ref object 0.107
 Closures 11.914
 Methods 1.298999999999999
```

**Comment by ringabout** [comment]
True, but in my case, there are no significant differences with threads on/off in refc after my PR, which indicates there shall be no regressions. My result is `Methods 0.1920000000000002` vs `Methods 0.2000000000000002`.

In your case, it is `Methods 0.3260000000000001` vs `Methods 1.298999999999999`.

That's the contradiction.

**Comment by Araq** [comment]
The refc code path is not changed at all. No idea what you measure. :P

**Comment by rockcavera** [comment]
In fact, I must have done something wrong in the last test. I recompiled and re-tested and indeed there is no regression for refc.

```
before this PR -d:release --mm:refc --threads:off
 Proc with ref object 0.108
 Closures 2.265
 Methods 0.359

after this PR -d:release --mm:refc --threads:off
 Proc with ref object 0.108
 Closures 2.353
 Methods 0.3599999999999999

before this PR -d:release --mm:refc --threads:on
 Proc with ref object 0.107
 Closures 4.292999999999999
 Methods 0.4500000000000002

after this PR -d:release --mm:refc --threads:on
 Proc with ref object 0.107
 Closures 4.306
 Methods 0.4340000000000002
```

**Pull request merged**